### PR TITLE
nixos/ipfs: Only set ReadWritePaths when hardened

### PR DESCRIPTION
**Pull request by max-privatevoid**
###### Description of changes

Setting ReadWritePaths kills mount propagation and unconditionally breaks FUSE.

Workaround:
```
systemd.services.ipfs.serviceConfig.ReadWritePaths = lib.mkForce [ ];
```

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by marius851000** [comment]
I can confirm that applying ``systemd.services.ipfs.serviceConfig.ReadWritePaths = lib.mkForce [ ];`` to my configuration fix my broken ``/ipfs`` mount. (broken as in it does (correctly) refuse to list subdir of ``/ipfs``, but also of every other path that should have worked).

File: nixos/modules/services/network-filesystems/ipfs.nix [hunk 1]
```diff
@@ -280,7 +280,7 @@ in
         User = cfg.user;
         Group = cfg.group;
         StateDirectory = "";
-        ReadWritePaths = [ "" cfg.dataDir ];
+        ReadWritePaths = optionals cfg.autoMount [ "" cfg.dataDir ];
```
Comment (Luflosi) [inline nixos/modules/services/network-filesystems/ipfs.nix:283]:
```suggestion
        ReadWritePaths = optionals (!cfg.autoMount) [ "" cfg.dataDir ];
```
This condition needs to be inverted.

**Review by Luflosi — APPROVED** [review]
Test runs as expected, LGTM.

File: nixos/tests/ipfs.nix [hunk 2]
```diff
@@ -40,5 +48,12 @@ import ./make-test-python.nix ({ pkgs, ...} : {
     # Test if setting dataDir works properly with the hardened systemd unit
     machine.succeed("test -e /mnt/ipfs/config")
     machine.succeed("test ! -e /var/lib/ipfs/")
+
+    # Test FUSE mountpoint
+    ipfs_hash = fuse.succeed(
+        "echo fnord3 | ipfs --api /ip4/127.0.0.1/tcp/2324 add | awk '{ print $2 }'"
```
Comment (Luflosi) [inline nixos/tests/ipfs.nix:54]:
```suggestion
        "echo fnord3 | ipfs --api /ip4/127.0.0.1/tcp/2324 add --quieter"
```

Comment (Luflosi) [reply nixos/tests/ipfs.nix:54]: Minor improvement.

**Review by Luflosi — APPROVED** [review]

**Review by fufexan — APPROVED** [review]
@Artturin can you merge this?

**Review by McSinyx — APPROVED** [review]
I also confirm that this fixes the mount points.

**Pull request merged**